orchagent/fabricportsorch.cpp

```diff
@@ -0,0 +1,268 @@
+
+    m_isQueueStatsGenerated = true;
+}
+
+void FabricPortsOrch::updateFabricPortState()
+{
+    if (!m_getFabricPortListDone) return;
+
+    SWSS_LOG_ENTER();
+
+    sai_status_t status;
+    sai_attribute_t attr;
+
+    time_t now;
+    struct timespec time_now;
+    if (clock_gettime(CLOCK_MONOTONIC, &time_now) < 0)
+    {
+        return;
+    }
+    now = time_now.tv_sec;
+
+    for (auto p : m_fabricLanePortMap)
+    {
+        int lane = p.first;
+        sai_object_id_t port = p.second;
+
+        string key = "PORT" + to_string(lane);
+        std::vector<FieldValueTuple> values;
+        uint32_t remote_peer;
+        uint32_t remote_port;
+
+        attr.id = SAI_PORT_ATTR_FABRIC_ATTACHED;
+        status = sai_port_api->get_port_attribute(port, 1, &attr);
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            // Port may not be ready for query
+            SWSS_LOG_ERROR("Failed to get fabric port (%d) status, rv:%d", lane, status);
+            return;
+        }
+
+        if (m_portStatus.find(lane) != m_portStatus.end() &&
+            m_portStatus[lane] && !attr.value.booldata)
+        {
+            m_portDownCount[lane] ++;
+            m_portDownSeenLastTime[lane] = now;
+        }
+        m_portStatus[lane] = attr.value.booldata;
+
+        if (m_portStatus[lane])
+        {
+            attr.id = SAI_PORT_ATTR_FABRIC_ATTACHED_SWITCH_ID;
+            status = sai_port_api->get_port_attribute(port, 1, &attr);
+            if (status != SAI_STATUS_SUCCESS)
+            {
+                throw runtime_error("FabricPortsOrch get remote id failure");
+            }
+            remote_peer = attr.value.u32;
+
+            attr.id = SAI_PORT_ATTR_FABRIC_ATTACHED_PORT_INDEX;
+            status = sai_port_api->get_port_attribute(port, 1, &attr);
+            if (status != SAI_STATUS_SUCCESS)
+            {
+                throw runtime_error("FabricPortsOrch get remote port index failure");
+            }
+            remote_port = attr.value.u32;
+        }
+
+        values.emplace_back("STATUS", m_portStatus[lane] ? "up" : "down");
+        if (m_portStatus[lane])
+        {
+            values.emplace_back("REMOTE_MOD", to_string(remote_peer));
+            values.emplace_back("REMOTE_PORT", to_string(remote_port));
+        }
+        if (m_portDownCount[lane] > 0)
+        {
+            values.emplace_back("PORT_DOWN_COUNT", to_string(m_portDownCount[lane]));
+            values.emplace_back("PORT_DOWN_SEEN_LAST_TIME",
+                                to_string(m_portDownSeenLastTime[lane]));
+        }
+        m_stateTable->set(key, values);
+    }
+}
+
+void FabricPortsOrch::doTask()
+{
+}
+
+void FabricPortsOrch::doTask(Consumer &consumer)
+{
+}
+
+void FabricPortsOrch::doTask(swss::SelectableTimer &timer)
+{
+    SWSS_LOG_ENTER();
+
+    if (!m_getFabricPortListDone)
+    {
+        getFabricPortList();
+    }
+
+    if (m_getFabricPortListDone)
+    {
+        updateFabricPortState();
```

Comment:
can we add link to the PR description where these State DB fields are define. Also looks like current VS test case added as part of this do not cover state db changes.

Reply:
Sure. I updated the description with the link.

The VS test already verifies the number of fabric ports in `STATE_DB` which is 16 fabric ports created for voq asic. Please look at the test part:
```
        # Verify that fabric ports are monitored in STATE_DB
        if metadata.get('switch_type', 'npu') in ['voq', 'fabric']:
            self.get_state_db()
            self.state_db.wait_for_n_keys("FABRIC_PORT_TABLE", 16)
```
Below is what we see from a container of a linecard created by a vs test.
```
root@a5959e6e0a4a:/# sonic-db-cli STATE_DB HGETALL 'FABRIC_PORT_TABLE|PORT10'
{'STATUS': 'down'}
```
In addition to the change in the test that verifies `STATE_DB`, https://github.com/Azure/sonic-sairedis/pull/737 merged to vslib to create fabric ports for voq asics. 

Note that we don't test fabric asics because it will require significant changes in the tests - all tests now assume testing asic is NPU with front panel ports.